File: code/java/Exercises.java
import java.io.BufferedReader;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.IOException;
import java.util.List;
import java.util.Optional;
import java.util.function.Predicate;

public class Exercises {

    public static Optional<String> firstThenLowerCase(List<String> xs, Predicate<String> p) {
        for (String s : xs) {
            if (p.test(s)) return Optional.of(s.toLowerCase());
        }
        return Optional.empty();
    }

    public static Say say() { return new Say(""); }
    public static Say say(String first) { return new Say(first); }

    public static final class Say {
        private final String built;
        private Say(String built) { this.built = built; }
        public Say and(String next) { return new Say(built + " " + next); }
        public String phrase() { return built; }
    }

    public static int meaningfulLineCount(String path) throws IOException {
        try (BufferedReader br = new BufferedReader(new FileReader(path))) {
            int count = 0;
            String line;
            while ((line = br.readLine()) != null) {
                String t = line.trim();
                if (t.isEmpty()) continue;
                if (t.startsWith("#")) continue;
                if (t.startsWith("//")) continue;
                count++;
            }
            return count;
        } catch (FileNotFoundException e) {
            FileNotFoundException x = new FileNotFoundException("No such file: " + path);
            x.initCause(e);
            throw x;
        }
    }
}

final class Quaternion {
    private final double a, b, c, d; 

    static final Quaternion ZERO = new Quaternion(0,0,0,0);
    static final Quaternion I    = new Quaternion(0,1,0,0);
    static final Quaternion J    = new Quaternion(0,0,1,0);
    static final Quaternion K    = new Quaternion(0,0,0,1);

    Quaternion(double a, double b, double c, double d) {
        if (Double.isNaN(a) || Double.isNaN(b) || Double.isNaN(c) || Double.isNaN(d))
            throw new IllegalArgumentException("Coefficients cannot be NaN");
        this.a = a; this.b = b; this.c = c; this.d = d;
    }

    double a() { return a; }
    double b() { return b; }
    double c() { return c; }
    double d() { return d; }

    List<Double> coefficients() { return List.of(a, b, c, d); }

    Quaternion plus(Quaternion q) {
        return new Quaternion(a + q.a, b + q.b, c + q.c, d + q.d);
    }

    Quaternion times(Quaternion q) {
        double e = q.a, f = q.b, g = q.c, h = q.d;
        double ra = a*e - b*f - c*g - d*h;
        double rb = a*f + b*e + c*h - d*g;
        double rc = a*g - b*h + c*e + d*f;
        double rd = a*h + b*g - c*f + d*e;
        return new Quaternion(ra, rb, rc, rd);
    }

    Quaternion conjugate() { return new Quaternion(a, -b, -c, -d); }

    @Override
    public boolean equals(Object o) {
        if (!(o instanceof Quaternion q)) return false;
        return Double.compare(a, q.a) == 0 &&
               Double.compare(b, q.b) == 0 &&
               Double.compare(c, q.c) == 0 &&
               Double.compare(d, q.d) == 0;
    }

    @Override
    public int hashCode() {
        return java.util.Objects.hash(a,b,c,d);
    }

    @Override
    public String toString() {
        if (a == 0.0 && b == 0.0 && c == 0.0 && d == 0.0) return "0";

        StringBuilder s = new StringBuilder();

        if (a != 0.0) s.append(Double.toString(a));

        appendImag(s, c, 'j');
        appendImag(s, b, 'i');
        appendImag(s, d, 'k');

        return s.length() == 0 ? "0" : s.toString();
    }

    private static void appendImag(StringBuilder s, double coef, char unit) {
        if (coef == 0.0) return;
        String sign = coef < 0 ? "-" : (s.length() == 0 ? "" : "+");
        double mag = Math.abs(coef);
        s.append(sign);
        if (mag == 1.0) s.append(unit);
        else s.append(Double.toString(mag)).append(unit);
    }
}

interface BinarySearchTree {
    int size();
    boolean contains(String s);
    BinarySearchTree insert(String s);
}

final class Empty implements BinarySearchTree {
    @Override public int size() { return 0; }
    @Override public boolean contains(String s) { return false; }
    @Override public BinarySearchTree insert(String s) { return new Node(s, this, this); }
    @Override public String toString() { return "()"; }
}

final class Node implements BinarySearchTree {
    private final String value;
    private final BinarySearchTree left, right;

    Node(String value, BinarySearchTree left, BinarySearchTree right) {
        this.value = value; this.left = left; this.right = right;
    }

    @Override public int size() { return 1 + left.size() + right.size(); }

    @Override public boolean contains(String s) {
        int cmp = s.compareTo(value);
        if (cmp == 0) return true;
        return (cmp < 0) ? left.contains(s) : right.contains(s);
    }

    @Override public BinarySearchTree insert(String s) {
        int cmp = s.compareTo(value);
        if (cmp == 0) return this; 
        if (cmp < 0)  return new Node(value, left.insert(s), right);
        else          return new Node(value, left, right.insert(s));
    }

    @Override public String toString() {
        String l = (left instanceof Empty) ? "" : left.toString();
        String r = (right instanceof Empty) ? "" : right.toString();
        return "(" + l + value + r + ")";
    }
}

File: code/kotlin/Exercises.kt
import java.io.File
import java.io.FileNotFoundException
import java.io.IOException

fun firstThenLowerCase(xs: List<String>, p: (String) -> Boolean): String? {
    for (s in xs) if (p(s)) return s.lowercase()
    return null
}

fun say(): Say = Say("")
fun say(first: String): Say = Say(first)

class Say(private val built: String) {
    val phrase: String get() = built
    fun and(next: String): Say = Say("$built $next")
}

@Throws(IOException::class)
fun meaningfulLineCount(path: String): Long {
    try {
        val file = File(path)
        if (!file.exists()) throw FileNotFoundException("No such file: $path")
        return file.readLines().count {
            val t = it.trim()
            t.isNotEmpty() && !t.startsWith("#") && !t.startsWith("//")
        }.toLong()
    } catch (e: FileNotFoundException) {
        throw e
    }
}

data class Quaternion(val a: Double, val b: Double, val c: Double, val d: Double) {
    init {
        if (listOf(a, b, c, d).any { it.isNaN() })
            throw IllegalArgumentException("Coefficients cannot be NaN")
    }

    fun coefficients(): List<Double> = listOf(a, b, c, d)
    fun conjugate(): Quaternion = Quaternion(a, -b, -c, -d)
    operator fun plus(q: Quaternion) = Quaternion(a + q.a, b + q.b, c + q.c, d + q.d)
    operator fun times(q: Quaternion): Quaternion {
        val e = q.a; val f = q.b; val g = q.c; val h = q.d
        return Quaternion(
            a * e - b * f - c * g - d * h,
            a * f + b * e + c * h - d * g,
            a * g - b * h + c * e + d * f,
            a * h + b * g - c * f + d * e
        )
    }

    override fun toString(): String {
        if (a == 0.0 && b == 0.0 && c == 0.0 && d == 0.0) return "0"
        val s = StringBuilder()
        if (a != 0.0) s.append(a)
        appendImag(s, b, 'i')
        appendImag(s, c, 'j')
        appendImag(s, d, 'k')
        return if (s.isEmpty()) "0" else s.toString()
    }

    private fun appendImag(sb: StringBuilder, coef: Double, unit: Char) {
        if (coef == 0.0) return
        val sign = if (coef < 0) "-" else if (sb.isEmpty()) "" else "+"
        val mag = kotlin.math.abs(coef)
        sb.append(sign)
        if (mag == 1.0) sb.append(unit)
        else sb.append(mag).append(unit)
    }

    companion object {
        val ZERO = Quaternion(0.0, 0.0, 0.0, 0.0)
        val I = Quaternion(0.0, 1.0, 0.0, 0.0)
        val J = Quaternion(0.0, 0.0, 1.0, 0.0)
        val K = Quaternion(0.0, 0.0, 0.0, 1.0)
    }
}

sealed interface BinarySearchTree {
    fun size(): Int
    fun contains(s: String): Boolean
    fun insert(s: String): BinarySearchTree
    companion object { val Empty: BinarySearchTree = EmptyTree }
}

object EmptyTree : BinarySearchTree {
    override fun size() = 0
    override fun contains(s: String) = false
    override fun insert(s: String): BinarySearchTree = Node(s, this, this)
    override fun toString() = "()"
}

data class Node(val value: String, val left: BinarySearchTree, val right: BinarySearchTree) : BinarySearchTree {
    override fun size(): Int = 1 + left.size() + right.size()
    override fun contains(s: String): Boolean = when {
        s == value -> true
        s < value -> left.contains(s)
        else -> right.contains(s)
    }
    override fun insert(s: String): BinarySearchTree = when {
        s == value -> this
        s < value -> Node(value, left.insert(s), right)
        else -> Node(value, left, right.insert(s))
    }
    override fun toString(): String {
        val l = if (left == BinarySearchTree.Empty) "" else left.toString()
        val r = if (right == BinarySearchTree.Empty) "" else right.toString()
        return "($l$value$r)"
    }
}
